R/data.table.R

```diff
@@ -1942,9 +1942,42 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
       X[[j]] = xj
     }
   }
-  X = unlist(X, recursive = FALSE, use.names = FALSE)
-  dim(X) <- c(n, length(X)/n)
-  dimnames(X) <- list(rownames.value, unlist(collabs, use.names = FALSE))
+  
+  # If any columns of x are multi-column, which can only occur if x is
+  # constructed incorrectly (see test 2074.07), use the old non-C
+  # method of converting to a matrix.
+  col.classes = sapply(X, class)
+  if (any(col.classes %in% c("matrix", "data.frame", "data.table"))) { 
+    X = unlist(X, recursive = FALSE, use.names = FALSE)
+    dim(X) = c(n, length(X)/n)
+    dimnames(X) = list(rownames.value, unlist(collabs, use.names = FALSE))
+    return(X)
+  } 
+  
+  # convert columns to common class before handing over to C
+  class.order = c("logical"=1, "integer"=2, "numeric"=3, "complex"=4,
+                  "character"=5, "raw"=6, "list"=7)
+  if (length(unique(col.classes)) > 1) { 
```

Comment:
please use explicit integers where appropriate 😬

Reply:
Good catch. fixed in latest commit